src/hooks/useI18n.js

```diff
@@ -1,4 +1,17 @@
-import { useContext } from 'preact/hooks'
-import { I18nContext } from 'contexts'
-
-export const useI18n = () => useContext(I18nContext)
+import { useContext } from 'preact/hooks'
+import { I18nContext } from 'contexts'
+
+export const useI18n = () => {
+  const i18n = useContext(I18nContext)
+
+  i18n.getTranslation = (locale, key, ...parameters) => {

```

Comment:
Modifying an external object like that is generally not recommended but I think you're on to something. What about returning a function instead of an object, and depending on whether `useI18n()` is called with a `lang` argument, the function could switch the locale.

A positive side-effect would be that usage everywhere would look like `i18n('softkey-blah')` instead of `i18n.i18n('softkey-blah')`

Reply:
Hmm I like where this is going

Reply:
`i18n()` is a good idea! The only concern I have is where the lang parameter stay? The arguments are message, arguments and lang.

Reply:
or maybe `useI18n()` returns two method?

one is `i18n( message, ...parameters )`

another one with lang `i18nGetTranslation( lang, message, ...parameters )`?



OR

we have two versions of `useI18n()`

`useI18n()` and `useExtendedI18n()` to provide two kinds of i18n instance

Reply:
after several trying, we can't use `useI18n( lang )` as there is only one instance, so unable to share between `useI18n()` and `useI18n( lang )`

I will continue try and see what else do we have

welcome your idea here 🍺

Reply:
I have not tried it but this is roughly what I meant:

```
export const useI18n = lang => {
	const banana = useContext(I18nContext)

	return (key, ...args) => {
		if (lang) {
			const locale = banana.locale
			banana.setLocale(lang)
			const msg = banana.i18n(key, args)
			banana.setLocale(locale)
			return msg
		}
		return banana.i18n(key, args)
	}
}

```

The shared instance is only modified inside the call to `i18n()` and the `lang` argument is kept in the closure so different calls of `useI18n()` have different languages. You should be able to do
```
const i18nAppLang = useI18n()
const i18nContentLang = useLang(article.contentLang)
```